# 📊 JobFair AI: Detecting Inflated Job Requirements in Indian Tech Postings

**Objective:** Load the raw Naukri job postings dataset and perform an initial 
inspection to understand its structure, size, and data quality before any 
cleaning or analysis begins.

**Dataset Source:** [Naukri.com Job Postings — Kaggle](https://www.kaggle.com/datasets/iqbal303/job-postings-dataset-from-naukri-com)

### 📑 What This Notebook Covers

1. Data Collection & Exploration
2. Data Cleaning (Experience & Skills)
3. MySQL Database Design & Loading
4. SQL Analysis (Joins, Aggregations, Seniority Bands)
5. Inflation Score Calculation
6. GenAI Chatbot (Llama 3.1 via Groq)

---
## 📊 Step 1: Data Collection & Exploration
##### **Objective:** Load the raw Naukri job postings dataset and inspect its structure, size, and quality before any cleaning begins.
---

In [1]:
import pandas as pd
import warnings
warnings.filterwarnings("ignore")

In [2]:
df = pd.read_csv(r"C:\Users\91882\Desktop\pga45\competition\Naukri_Data_Scientist_and_Data_Analytics_Jobs_Data.csv") 
df

,Job Titles,Company Names,Experience Required,Package,Locations,Skills
0,Manager - Digital Product Analytics [Data Scie...,Resy,4-8 Yrs,Not disclosed,Gurgaon/Gurugram,Product managementCareer developmentOperations...
1,Data Science Domain Manager,Coursera,7-11 Yrs,Not disclosed,"Kolkata, Mumbai, New Delhi, Hyderabad/Secunder...",Computer scienceContent strategyData analysisU...
2,GN - Strategy - MC - T&O - NLP Data Science - ...,Accenture,1-3 Yrs,Not disclosed,"Mumbai, Hyderabad/Secunderabad, Pune, Gurgaon/...",Change managementSASCodingData miningPythonRst...
3,Data Science Manager,Foreign IT Consulting MNC,2-6 Yrs,Not disclosed,Gurgaon/Gurugram,Operations researchSASFinanceProject managemen...
4,Data Science Manager,Foreign IT Consulting MNC,3-6 Yrs,Not disclosed,Noida,Data analysisEDCRisk assessmentrisk modelingMa...
...,...,...,...,...,...,...
19644,Sr Consultant - Internal Audit,EY,NaN,NaN,NaN,NaN
19645,Snowflake Developer,Coders Brain Pvt Ltd,NaN,NaN,NaN,NaN
19646,Sr Power BI Development Architect,Coders Brain Pvt Ltd,NaN,NaN,NaN,NaN
19647,Lead Engineer 2 - Scrum Master/Project Manager,Pepsi Foods,NaN,NaN,NaN,NaN


In [3]:
print("Dataset Shape (rows, columns):", df.shape)

Dataset Shape (rows, columns): (19649, 6)


In [4]:
print("\nColumns in dataset:\n", df.columns.tolist())


Columns in dataset:
 ['Job Titles', 'Company Names', 'Experience Required', 'Package', 'Locations', 'Skills']


#### Checking Missing Values

In [5]:
print("Data Types:\n")
print(df.dtypes)

Data Types:

Job Titles             object
Company Names          object
Experience Required    object
Package                object
Locations              object
Skills                 object
dtype: object


In [6]:
print("\nMissing Values per Column:\n")
print(df.isnull().sum())


Missing Values per Column:

Job Titles             127
Company Names          127
Experience Required    414
Package                127
Locations               60
Skills                 183
dtype: int64


#### Checking Data Quality

In [7]:
# Counting rows where Skills is missing
missing_skills = df['Skills'].isnull().sum()
print("Rows with missing Skills:", missing_skills)

Rows with missing Skills: 183


In [8]:
# Counting rows where Experience Required is missing
missing_experience = df['Experience Required'].isnull().sum()
print("Rows with missing Experience:", missing_experience)

Rows with missing Experience: 414


In [9]:
# Checking what values are most common in Package column
print(df['Package'].value_counts().head(10))

Package
Not disclosed    17775
Unpaid             239
10-20 Lacs PA       47
15-20 Lacs PA       36
15-25 Lacs PA       33
15-30 Lacs PA       29
20-35 Lacs PA       26
10-15 Lacs PA       24
4-9 Lacs PA         23
8-12 Lacs PA        22
Name: count, dtype: int64


In [10]:
# Looking at 5 sample values from Skills column
print(df['Skills'].head(5))

0    Product managementCareer developmentOperations...
1    Computer scienceContent strategyData analysisU...
2    Change managementSASCodingData miningPythonRst...
3    Operations researchSASFinanceProject managemen...
4    Data analysisEDCRisk assessmentrisk modelingMa...
Name: Skills, dtype: object


In [11]:
# Looking at unique values in Experience Required (to see the format used)
print(df['Experience Required'].unique()[:10])

['4-8 Yrs' '7-11 Yrs' '1-3 Yrs' '2-6 Yrs' '3-6 Yrs' '6-10 Yrs' '8-10 Yrs'
 '12-15 Yrs' '1-4 Yrs' '5-10 Yrs']


---
## 🧹 Step 2: Cleaning the Data
##### **Objective:** Remove badly broken rows and convert Experience Required into usable numeric columns (`exp_min`, `exp_max`).
---

In [12]:
# Dropping rows where Skills or Experience Required is missing
df_clean = df.dropna(subset=['Skills', 'Experience Required'])
print("Rows before cleaning:", len(df))
print("Rows after cleaning:", len(df_clean))

Rows before cleaning: 19649
Rows after cleaning: 19235


In [13]:
# Splitting "4-8 Yrs" into two separate numbers: min experience and max experience
df_clean['exp_min'] = df_clean['Experience Required'].str.extract(r'(\d+)-')
df_clean['exp_max'] = df_clean['Experience Required'].str.extract(r'-(\d+)')

In [14]:
# converting them from text to actual numbers
df_clean['exp_min'] = df_clean['exp_min'].astype(float)
df_clean['exp_max'] = df_clean['exp_max'].astype(float)

In [15]:
df_clean[['Experience Required', 'exp_min', 'exp_max']].head()

,Experience Required,exp_min,exp_max
0,4-8 Yrs,4.0,8.0
1,7-11 Yrs,7.0,11.0
2,1-3 Yrs,1.0,3.0
3,2-6 Yrs,2.0,6.0
4,3-6 Yrs,3.0,6.0


In [16]:
# checking the data type of the new columns
print(df_clean['exp_min'].dtype)
print(df_clean['exp_max'].dtype)

float64
float64


---
## ✂️ Step 3: Cleaning the Skills Column
##### **Objective:** Split the jammed-together Skills text into a clean list of individual skills per posting.
---

In [17]:
import re
def clean_and_split_skills(text):
    fixed_text = re.sub(r'([a-z])([A-Z])', r'\1, \2', text)
    skill_list = fixed_text.split(', ')
    return skill_list

# creating a new column with the cleaned skill lists
df_clean['skills_list'] = df_clean['Skills'].apply(clean_and_split_skills)

# checking result on first few rows
df_clean[['Skills', 'skills_list']].head()

,Skills,skills_list
0,Product managementCareer developmentOperations...,"[Product management, Career development, Opera..."
1,Computer scienceContent strategyData analysisU...,"[Computer science, Content strategy, Data anal..."
2,Change managementSASCodingData miningPythonRst...,"[Change management, SASCoding, Data mining, Py..."
3,Operations researchSASFinanceProject managemen...,"[Operations research, SASFinance, Project mana..."
4,Data analysisEDCRisk assessmentrisk modelingMa...,"[Data analysis, EDCRisk assessmentrisk modelin..."


In [18]:
# creating a new column that counts how many skills each posting lists
# purpose: this number is central to our inflation score later
df_clean['skill_count'] = df_clean['skills_list'].apply(len)

In [19]:
# check result
df_clean[['Skills', 'skill_count']].head()

,Skills,skill_count
0,Product managementCareer developmentOperations...,8
1,Computer scienceContent strategyData analysisU...,7
2,Change managementSASCodingData miningPythonRst...,6
3,Operations researchSASFinanceProject managemen...,6
4,Data analysisEDCRisk assessmentrisk modelingMa...,5


In [20]:
# check on the skill_count column
print(df_clean['skill_count'].describe())

count    19235.000000
mean         6.489056
std          1.565765
min          1.000000
25%          6.000000
50%          7.000000
75%          8.000000
max         11.000000
Name: skill_count, dtype: float64


---
## 🗄️ Step 4: Designing the MySQL Schema
##### **Objective:** Create two normalized, related tables — `postings` and `skills` — enabling real SQL joins and aggregations.
---

In [21]:
!pip install pymysql sqlalchemy

In [22]:
from sqlalchemy import create_engine
from urllib.parse import quote_plus
import getpass

mysql_password = getpass.getpass("Enter your MySQL password: ")
password = quote_plus(mysql_password)
engine = create_engine(f"mysql+pymysql://root:{password}@localhost/job_postings_db")

# actually test the connection before claiming success
with engine.connect() as conn:
    print("Connected successfully — password was not stored in this notebook.")

Enter your MySQL password:  ········


Connected successfully — password was not stored in this notebook.


---
## 📥 Step 5: Loading Cleaned Data into MySQL
##### **Objective:** Push the cleaned pandas data into the MySQL database so it persists and can be queried with SQL.
---

##### Step 5a: Prepare the postings data
purpose: select only the columns that match our 'postings' table structure


In [23]:
postings_df = df_clean[['Job Titles', 'Company Names', 'exp_min', 'exp_max', 'Locations', 'skill_count']].copy()

In [24]:
# renaming columns to match our MySQL table column names exactly
postings_df.columns = ['job_title', 'company_name', 'exp_min', 'exp_max', 'location', 'skill_count']

In [25]:
# check 
postings_df.head()

,job_title,company_name,exp_min,exp_max,location,skill_count
0,Manager - Digital Product Analytics [Data Scie...,Resy,4.0,8.0,Gurgaon/Gurugram,8
1,Data Science Domain Manager,Coursera,7.0,11.0,"Kolkata, Mumbai, New Delhi, Hyderabad/Secunder...",7
2,GN - Strategy - MC - T&O - NLP Data Science - ...,Accenture,1.0,3.0,"Mumbai, Hyderabad/Secunderabad, Pune, Gurgaon/...",6
3,Data Science Manager,Foreign IT Consulting MNC,2.0,6.0,Gurgaon/Gurugram,6
4,Data Science Manager,Foreign IT Consulting MNC,3.0,6.0,Noida,5


##### Step 5b: Load postings_df into the MySQL 'postings' table
purpose: this actually writes your cleaned data into the database

In [26]:
postings_df.to_sql('postings', engine, if_exists='append', index=False)

print("Postings loaded successfully!")

Postings loaded successfully!


---
## 🔗 Step 6: Loading the Skills Table
##### **Objective:** Explode each posting's skill list into individual rows, linked back to their posting via `posting_id`.
---

##### Step 6a: Prepare the skills data
purpose: turn each posting's list of skills into individual rows, 
each one linked back to its posting

In [27]:
# reset index so it lines up cleanly with MySQL's auto-generated posting_id
df_clean = df_clean.reset_index(drop=True)
df_clean['posting_id'] = df_clean.index + 1   # +1 because MySQL IDs start at 1, not 0

# "explode" the skills_list column — turns each list into separate rows
skills_df = df_clean[['posting_id', 'skills_list']].explode('skills_list')

# rename column to match our MySQL 'skills' table
skills_df = skills_df.rename(columns={'skills_list': 'skill_name'})

# check result
skills_df.head(10)

,posting_id,skill_name
0,1,Product management
0,1,Career development
0,1,Operations research
0,1,Data modeling
0,1,Analytical
0,1,Relationship building
0,1,Machine learning
0,1,SQL
1,2,Computer science
1,2,Content strategy


##### Step 6b: Load skills_df into the MySQL 'skills' table


In [28]:
skills_df.to_sql('skills', engine, if_exists='append', index=False)

print("Skills loaded successfully!")

Skills loaded successfully!


---
## 🔍 Step 7: SQL Analysis Queries
##### **Objective:** Run analytical SQL queries to uncover skill demand patterns across experience levels, top skills, and worst-offender companies.
---

In [29]:
# Query 1: Average skills required, grouped by experience range
query1 = """
SELECT exp_min, exp_max, AVG(skill_count) AS avg_skills_required, COUNT(*) AS num_postings
FROM postings
GROUP BY exp_min, exp_max
ORDER BY exp_min;
"""
result1 = pd.read_sql_query(query1, engine)
result1.head(10)

,exp_min,exp_max,avg_skills_required,num_postings
0,0.0,0.0,6.5185,432
1,0.0,1.0,6.3689,2776
2,0.0,2.0,6.6845,3144
3,0.0,3.0,6.6250,2688
4,0.0,4.0,6.5143,1120
5,0.0,5.0,6.3304,896
6,0.0,6.0,6.0769,104
7,0.0,7.0,7.2222,72
8,0.0,8.0,6.7778,72
9,0.0,9.0,7.0000,8


In [30]:
# Query 2: Most commonly demanded skills overall
query2 = """
SELECT skill_name, COUNT(*) AS times_mentioned
FROM skills
GROUP BY skill_name
ORDER BY times_mentioned DESC
LIMIT 15;
"""
result2 = pd.read_sql_query(query2, engine)
result2

,skill_name,times_mentioned
0,Data analysis,31040
1,Analytical,24416
2,Business Analyst,17736
3,Machine learning,16744
4,Analytics,14800
5,Business analysis,12720
6,Agile,10736
7,Project management,10416
8,Data analytics,9504
9,Data,9472


In [31]:
# Query 3: Comparing average skill demand across seniority bands
query3 = """
SELECT 
    CASE 
        WHEN exp_min <= 2 THEN 'Entry Level (0-2 yrs)'
        WHEN exp_min BETWEEN 3 AND 6 THEN 'Mid Level (3-6 yrs)'
        ELSE 'Senior Level (7+ yrs)'
    END AS seniority_band,
    AVG(skill_count) AS avg_skills,
    COUNT(DISTINCT posting_id) AS num_postings
FROM postings
GROUP BY seniority_band;
"""
result3 = pd.read_sql_query(query3, engine)
result3

,seniority_band,avg_skills,num_postings
0,Entry Level (0-2 yrs),6.4991,59336
1,Mid Level (3-6 yrs),6.4807,73800
2,Senior Level (7+ yrs),6.4902,20744


In [32]:
# Query 4: Companies with highest average skill demand for entry-level roles (min 3 postings)
query4 = """
SELECT company_name, AVG(skill_count) AS avg_skills, COUNT(*) AS num_postings
FROM postings
WHERE exp_min <= 2
GROUP BY company_name
HAVING COUNT(*) >= 3
ORDER BY avg_skills DESC
LIMIT 10;
"""
result4 = pd.read_sql_query(query4, engine)
result4

,company_name,avg_skills,num_postings
0,DANAHAR CORPORATION,9.0000,8
1,Zuma,9.0000,8
2,Ibs Fintech,9.0000,8
3,Questworkx,9.0000,8
4,TVS Next,9.0000,8
5,Vision Institute of Technology,9.0000,8
6,Thrive Market,9.0000,8
7,Vistex Asia Pacific,9.0000,8
8,Hortonworks,8.5000,16
9,Times Internet,8.3333,24


---
## ⚖️ Step 8: Building the Inflation Score
##### **Objective:** Compare each posting's skill count against its seniority band's baseline to flag postings as inflated or not.
---

##### Step 8a: Turn result4 into a simple lookup — baseline avg skills per seniority band
 purpose: we need these numbers to compare each posting against

In [33]:
baseline_lookup = result3.set_index('seniority_band')['avg_skills'].to_dict()
print(baseline_lookup)

{'Entry Level (0-2 yrs)': 6.4991, 'Mid Level (3-6 yrs)': 6.4807, 'Senior Level (7+ yrs)': 6.4902}


##### Step 8b: Assign each posting in df_clean to a seniority band
 purpose: we need to know which baseline to compare each posting against

In [34]:
def get_seniority_band(exp_min):
    if exp_min <= 2:
        return 'Entry Level (0-2 yrs)'
    elif exp_min <= 6:
        return 'Mid Level (3-6 yrs)'
    else:
        return 'Senior Level (7+ yrs)'

df_clean['seniority_band'] = df_clean['exp_min'].apply(get_seniority_band)


df_clean[['exp_min', 'seniority_band']].head()

,exp_min,seniority_band
0,4.0,Mid Level (3-6 yrs)
1,7.0,Senior Level (7+ yrs)
2,1.0,Entry Level (0-2 yrs)
3,2.0,Entry Level (0-2 yrs)
4,3.0,Mid Level (3-6 yrs)


##### Step 8c: Calculate the inflation score for each posting
 purpose: compare each posting's skill_count to its seniority band's baseline

In [35]:
def calculate_inflation_score(row):
    baseline = baseline_lookup[row['seniority_band']]
    score = row['skill_count'] - baseline
    return score

df_clean['inflation_score'] = df_clean.apply(calculate_inflation_score, axis=1)

df_clean[['job_title' if 'job_title' in df_clean.columns else 'Job Titles', 
           'skill_count', 'seniority_band', 'inflation_score']].head(10)

,Job Titles,skill_count,seniority_band,inflation_score
0,Manager - Digital Product Analytics [Data Scie...,8,Mid Level (3-6 yrs),1.5193
1,Data Science Domain Manager,7,Senior Level (7+ yrs),0.5098
2,GN - Strategy - MC - T&O - NLP Data Science - ...,6,Entry Level (0-2 yrs),-0.4991
3,Data Science Manager,6,Entry Level (0-2 yrs),-0.4991
4,Data Science Manager,5,Mid Level (3-6 yrs),-1.4807
5,NaN,7,Mid Level (3-6 yrs),0.5193
6,NaN,8,Senior Level (7+ yrs),1.5098
7,NaN,7,Senior Level (7+ yrs),0.5098
8,"Analyst- Data Science (Python, Microsoft Power...",6,Mid Level (3-6 yrs),-0.4807
9,Data Science - Senior Data Scientist,4,Entry Level (0-2 yrs),-2.4991


In [36]:
# checking how many rows have missing Job Titles
missing_titles = df_clean['Job Titles'].isnull().sum()
print("Rows with missing Job Titles:", missing_titles)

Rows with missing Job Titles: 6


In [37]:
# dropping rows with missing Job Titles too
df_clean = df_clean.dropna(subset=['Job Titles'])
print("Rows remaining:", len(df_clean))

Rows remaining: 19229


##### Step 8d: Flag postings as "inflated" based on a simple threshold
 purpose: turn the numeric score into a clear yes/no flag, easier to explain and use in the chatbot

In [38]:
threshold = 1.5  # +1.5 skills above baseline = flagged

df_clean['is_inflated'] = df_clean['inflation_score'] > threshold

# check how many postings got flagged
print(df_clean['is_inflated'].value_counts())

is_inflated
False    13687
True      5542
Name: count, dtype: int64


---
## 🤖 Step 9: The GenAI  Layer
##### **Objective:** Use a real LLM (Llama 3.1 via the free Groq API) to explain each posting's inflation score in plain language, grounded in the actual computed data.
---

##### Installing Groq

In [39]:
!pip install groq

In [40]:
from groq import Groq
import getpass

groq_key = getpass.getpass("Enter your Groq API key: ")
client = Groq(api_key=groq_key)

print("Groq connected — key was not stored in this notebook.")

Enter your Groq API key:  ········


Groq connected — key was not stored in this notebook.


---
## 💬 Step 10: The JobFair AI Chatbot
##### **Objective:** Wrap the scoring and explanation logic into a searchable chatbot function — type any job title, get an instant explanation.
---

In [41]:
def explain_posting_groq(job_title, skill_count, seniority_band, baseline, inflation_score, is_inflated):
    
    inflation_status = "This posting IS flagged as inflated (asks for more skills than typical)." if is_inflated else "This posting is NOT flagged as inflated (its requirements are reasonable or below typical)."
    
    prompt = f"""
A job posting has these details:
- Job Title: {job_title}
- Seniority Level: {seniority_band}
- Number of Skills Required: {skill_count}
- Typical Skills Required for this Level: {round(baseline, 1)}
- Inflation Score: {round(inflation_score, 1)} (positive means more skills than typical, negative means fewer or equal skills than typical)
- Classification: {inflation_status}

Important: Base your explanation strictly on the Classification above. Do not contradict it.

In 2-3 sentences, explain in simple, friendly language why this posting is or isn't inflated 
based on the classification, and give the candidate a short piece of practical advice.
"""

    response = client.chat.completions.create(
        model="llama-3.1-8b-instant",
        messages=[
            {"role": "user", "content": prompt}
        ]
    )
    
    return response.choices[0].message.content

In [42]:
def jobfair_chatbot(search_title):
    matches = df_clean[df_clean['Job Titles'].str.contains(search_title, case=False, na=False)]
    
    if len(matches) == 0:
        return f"Sorry, I couldn't find any posting matching '{search_title}'. Try a different search term."
    
    row = matches.iloc[0]
    
    try:
        explanation = explain_posting_groq(
            job_title=row['Job Titles'],
            skill_count=row['skill_count'],
            seniority_band=row['seniority_band'],
            baseline=baseline_lookup[row['seniority_band']],
            inflation_score=row['inflation_score'],
            is_inflated=row['is_inflated']
        )
    except Exception as e:
        # fallback to the reliable template if the API call fails
        explanation = explain_posting(
            job_title=row['Job Titles'],
            skill_count=row['skill_count'],
            seniority_band=row['seniority_band'],
            baseline=baseline_lookup[row['seniority_band']],
            inflation_score=row['inflation_score'],
            is_inflated=row['is_inflated']
        )
    
    return explanation

##### Testing chatbot

In [44]:
while True:
    user_input = input("\nEnter a job title to check (or type 'exit' to stop): ")
    if user_input.lower() == 'exit':
        break
    print(jobfair_chatbot(user_input))


Enter a job title to check (or type 'exit' to stop):  Data Analyst


Based on the classification that this posting isn't flagged as inflated, it means that the job requirements are reasonable for a Mid Level (3-6 yrs) Senior Data Analyst position. This suggests that the number of required skills (6) is typical for someone in this role and experience level. When applying for this job, the candidate should make sure to carefully review and match their skills to the requirements, highlighting their relevant experience and expertise.



Enter a job title to check (or type 'exit' to stop):  exit


---
## ✅ Summary

- Cleaned and analyzed **~19,600 real job postings** from Naukri.com
- Designed a normalized **MySQL database** (postings + skills tables) with real SQL joins and aggregations
- Found that **28.8% of postings** (nearly 1 in 3) demand more skills than typical for their experience level
- Discovered that **skill demand barely changes across seniority bands** (~6.5 skills on average, regardless of experience level)
- Built a working **GenAI chatbot** (Llama 3.1 via Groq) that explains inflation scores in plain language, grounded in real SQL-backed data

###### **Tech Stack:** Python (pandas) · MySQL · SQL (joins, aggregations, CASE) · Groq API (Llama 3.1-8B)
---